### DISCUSSION TITLE
F# interop - Passing Immutable Collections & F# Collections into F# Functions from C#

### DISCUSSION THREAD

**Discussion by andrew-bauer-calibrated** [dotnet/csharplang#7833] [Q&A] [answered]
I have noticed that when I run F# functions in C# that return collections (like a List), they are returned as an FSharp variation (which is not a problem), however when I pass a collection into an F# function as a parameter, I can only pass in a mutable version.

Is there a way to pass in either the FSharp or Immutable variant? I know that I can just run ToList() on pass in, however I don't want to open the list up to possibly being mutated if it is already in an immutable format.

**Comment by huoyaoyuan** [comment, marked as answer]
> they are returned as an FSharp variation (which is not a problem

This is not precise. The F# collection syntax provides F# collections. There are no limitations to use standard collections in F#.

> Is there a way to pass in either the FSharp or Immutable variant? I know that I can just run ToList() on pass in, however I don't want to open the list up to possibly being mutated if it is already in an immutable format.

You can just declare the return type as F# collection. It will be seen as `FSharp.Collections.List<T>`. For immutability, you can use `System.Collections.Immutable.ImmutableArray<T>` and others.

**Reply by andrew-bauer-calibrated** [reply]
Thanks for the quick response.
